From 39128328a339648ce09764cd4f59fe527448cacf Mon Sep 17 00:00:00 2001
From: Reuben D'Netto <rdnetto@gmail.com>
Date: Fri, 12 Aug 2016 18:31:56 +1000
Subject: [PATCH 1/2] Added support for window icons (_NET_WM_ICON property)

Rebase of patch originally by Marius Muja <mariusm@cs.ubc.ca>
---
 common.mk                 |  7 ++++
 include/atoms_rest.xmacro |  1 +
 include/data.h            |  5 +++
 include/window.h          |  9 +++++
 src/manage.c              | 11 ++++++
 src/render.c              |  6 ++++
 src/window.c              | 85 +++++++++++++++++++++++++++++++++++++++++++++++
 src/x.c                   | 78 +++++++++++++++++++++++++++++++++++++++++++
 8 files changed, 202 insertions(+)

diff --git a/common.mk b/common.mk
index 4fe8f2b..ffd2af2 100644
--- a/common.mk
+++ b/common.mk
@@ -1,5 +1,6 @@
 UNAME=$(shell uname)
 DEBUG=1
+USE_ICONS=1
 ASAN=0
 INSTALL=install
 LN=ln
@@ -65,6 +66,9 @@ I3_CPPFLAGS += -DMINOR_VERSION=${MINOR_VERSION}
 I3_CPPFLAGS += -DPATCH_VERSION=${PATCH_VERSION}
 I3_CPPFLAGS += -DSYSCONFDIR=\"${SYSCONFDIR}\"
 I3_CPPFLAGS += -DI3__FILE__=__FILE__
+ifeq ($(USE_ICONS),1)
+I3_CPPFLAGS += -DUSE_ICONS
+endif
 
 
 ## Libraries flags
@@ -114,6 +118,9 @@ XCB_WM_CFLAGS += $(call cflags_for_lib, xcb-randr)
 XCB_WM_LIBS   := $(call ldflags_for_lib, xcb-icccm,xcb-icccm)
 XCB_WM_LIBS   += $(call ldflags_for_lib, xcb-xinerama,xcb-xinerama)
 XCB_WM_LIBS   += $(call ldflags_for_lib, xcb-randr,xcb-randr)
+ifeq ($(USE_ICONS),1)
+XCB_WM_LIBS   += $(call ldflags_for_lib, xcb-image,xcb-image)
+endif
 
 # XCB cursor
 XCB_CURSOR_CFLAGS := $(call cflags_for_lib, xcb-cursor)
diff --git a/include/atoms_rest.xmacro b/include/atoms_rest.xmacro
index d461dc0..f32a7e1 100644
--- a/include/atoms_rest.xmacro
+++ b/include/atoms_rest.xmacro
@@ -1,6 +1,7 @@
 xmacro(_NET_WM_USER_TIME)
 xmacro(_NET_STARTUP_ID)
 xmacro(_NET_WORKAREA)
+xmacro(_NET_WM_ICON)
 xmacro(WM_PROTOCOLS)
 xmacro(WM_DELETE_WINDOW)
 xmacro(UTF8_STRING)
diff --git a/include/data.h b/include/data.h
index 3a059e7..cbc2ac6 100644
--- a/include/data.h
+++ b/include/data.h
@@ -426,6 +426,11 @@ struct Window {
 
     /* aspect ratio from WM_NORMAL_HINTS (MPlayer uses this for example) */
     double aspect_ratio;
+
+#ifdef USE_ICONS
+    /** Window icon, array of size 16x16 containing the ARGB pixels */
+    uint32_t* icon;
+#endif
 };
 
 /**
diff --git a/include/window.h b/include/window.h
index d0b97f1..1a0ddfc 100644
--- a/include/window.h
+++ b/include/window.h
@@ -74,6 +74,15 @@ void window_update_type(i3Window *window, xcb_get_property_reply_t *reply);
  */
 void window_update_hints(i3Window *win, xcb_get_property_reply_t *prop, bool *urgency_hint);
 
+
+#ifdef USE_ICONS
+/**
+ * Updates the _NET_WM_ICON
+ *
+ */
+void window_update_icon(i3Window *win, xcb_get_property_reply_t *prop);
+#endif
+
 /**
  * Updates the MOTIF_WM_HINTS. The container's border style should be set to
  * `motif_border_style' if border style is not BS_NORMAL.
diff --git a/src/manage.c b/src/manage.c
index 93272f1..5fc4679 100644
--- a/src/manage.c
+++ b/src/manage.c
@@ -92,6 +92,10 @@ void manage_window(xcb_window_t window, xcb_get_window_attributes_cookie_t cooki
         role_cookie, startup_id_cookie, wm_hints_cookie,
         wm_normal_hints_cookie, motif_wm_hints_cookie, wm_user_time_cookie, wm_desktop_cookie;
 
+#ifdef USE_ICONS
+    xcb_get_property_cookie_t wm_icon_cookie;
+#endif
+
     geomc = xcb_get_geometry(conn, d);
 
     /* Check if the window is mapped (it could be not mapped when intializing and
@@ -163,6 +167,9 @@ void manage_window(xcb_window_t window, xcb_get_window_attributes_cookie_t cooki
     motif_wm_hints_cookie = GET_PROPERTY(A__MOTIF_WM_HINTS, 5 * sizeof(uint64_t));
     wm_user_time_cookie = GET_PROPERTY(A__NET_WM_USER_TIME, UINT32_MAX);
     wm_desktop_cookie = GET_PROPERTY(A__NET_WM_DESKTOP, UINT32_MAX);
+#ifdef USE_ICONS
+    wm_icon_cookie = xcb_get_property_unchecked(conn, false, window, A__NET_WM_ICON, XCB_ATOM_CARDINAL, 0, UINT32_MAX);
+#endif
 
     DLOG("Managing window 0x%08x\n", window);
 
@@ -184,6 +191,10 @@ void manage_window(xcb_window_t window, xcb_get_window_attributes_cookie_t cooki
     window_update_hints(cwindow, xcb_get_property_reply(conn, wm_hints_cookie, NULL), &urgency_hint);
     border_style_t motif_border_style = BS_NORMAL;
     window_update_motif_hints(cwindow, xcb_get_property_reply(conn, motif_wm_hints_cookie, NULL), &motif_border_style);
+#ifdef USE_ICONS
+    window_update_icon(cwindow, xcb_get_property_reply(conn, wm_icon_cookie, NULL));
+#endif
+
     xcb_size_hints_t wm_size_hints;
     if (!xcb_icccm_get_wm_size_hints_reply(conn, wm_normal_hints_cookie, &wm_size_hints, NULL))
         memset(&wm_size_hints, '\0', sizeof(xcb_size_hints_t));
diff --git a/src/render.c b/src/render.c
index 9fa40f0..b9bfd4b 100644
--- a/src/render.c
+++ b/src/render.c
@@ -127,6 +127,12 @@ void render_con(Con *con, bool render_fullscreen) {
     /* find the height for the decorations */
     params.deco_height = render_deco_height();
 
+#ifdef USE_ICONS
+    /* minimum decoration height to allow icon to fit
+     * not actuuly required, icon would be cropped otherwise */
+    params.deco_height = params.deco_height<16 ? 16 : params.deco_height;
+#endif
+
     /* precalculate the sizes to be able to correct rounding errors */
     params.sizes = precalculate_sizes(con, &params);
 
diff --git a/src/window.c b/src/window.c
index d10811f..9084145 100644
--- a/src/window.c
+++ b/src/window.c
@@ -20,6 +20,9 @@ void window_free(i3Window *win) {
     FREE(win->class_instance);
     i3string_free(win->name);
     FREE(win->ran_assignments);
+#ifdef USE_ICONS
+    FREE(win->icon);
+#endif
     FREE(win);
 }
 
@@ -367,3 +370,85 @@ void window_update_motif_hints(i3Window *win, xcb_get_property_reply_t *prop, bo
 #undef MWM_DECOR_BORDER
 #undef MWM_DECOR_TITLE
 }
+
+
+#ifdef USE_ICONS
+/*
+ * Copy and resize icon if needed
+ */
+void copy_icon_with_resize(uint32_t *dst, int width, int height, uint32_t* src, int s_width, int s_height)
+{
+    int i, j;
+    if (width==s_width && height==s_height) {
+        /*  easy case, same dimensions, just copy data */
+        memcpy(dst, src, width*height*sizeof(uint32_t));
+    }
+    else {
+        uint32_t* row = src;
+        int xstep = s_width/width;
+        int ystep = s_height/height*s_width;
+
+        for(i=0; i < height; ++i) {
+            uint32_t* ptr = row;
+            for(j=0; j < width; ++j) {
+                *dst++ = *ptr;
+                ptr+=xstep;
+            }
+            row += ystep;
+        }
+    }
+}
+
+
+void window_update_icon(i3Window *win, xcb_get_property_reply_t *prop)
+{
+    uint32_t *data = NULL;
+    uint64_t len = 0;
+
+    if(!prop || prop->type != XCB_ATOM_CARDINAL || prop->format != 32) {
+        DLOG("_NET_WM_ICON is not set\n");
+        FREE(prop);
+        return;
+    }
+
+    uint32_t prop_value_len = xcb_get_property_value_length(prop);
+    uint32_t *prop_value = (uint32_t *) xcb_get_property_value(prop);
+
+    /* Find the number of icons in the reply. */
+    while(prop_value_len > (sizeof(uint32_t) * 2) && prop_value && prop_value[0] && prop_value[1])
+    {
+        /* Check that the property is as long as it should be (in bytes),
+           handling integer overflow. "+2" to handle the width and height
+           fields. */
+        const uint64_t crt_len = prop_value[0] * (uint64_t) prop_value[1];
+        const uint64_t expected_len = (crt_len + 2) * 4;
+        if(expected_len > prop_value_len)
+            break;
+
+        if (len==0 || (crt_len>=16*16 && crt_len<len)) {
+            len = crt_len;
+            data  = prop_value;
+        }
+        if (len==16*16) break; // found 16 pixels icon
+
+        /* Find pointer to next icon in the reply. */
+        prop_value_len -= expected_len;
+        prop_value = (uint32_t *) (((uint8_t *) prop_value) + expected_len);
+    }
+
+    if (!data ) {
+        DLOG("Could not get _NET_WM_ICON\n");
+        free(prop);
+        return;
+    }
+
+    LOG("Got _NET_WM_ICON of size: (%d,%d)\n", data[0], data[1]);
+
+    FREE(win->icon);
+    win->icon = malloc(16*16*sizeof(uint32_t));
+    copy_icon_with_resize(win->icon, 16, 16, data+2, data[0], data[1]);
+
+    free(prop);
+}
+#endif /* USE_ICONS */
+
diff --git a/src/x.c b/src/x.c
index f44bc37..906142a 100644
--- a/src/x.c
+++ b/src/x.c
@@ -11,6 +11,9 @@
  *
  */
 #include "all.h"
+#ifdef USE_ICONS
+#include <xcb/xcb_image.h>
+#endif
 
 xcb_window_t ewmh_window;
 
@@ -299,6 +302,44 @@ void x_window_kill(xcb_window_t window, kill_window_t kill_window) {
     free(event);
 }
 
+#ifdef USE_ICONS
+
+static inline uint32_t pixel_blend(uint32_t d, uint32_t s)
+{
+	const uint32_t a     = (s >> 24) + 1;
+
+	const uint32_t dstrb = d & 0xFF00FF;
+	const uint32_t dstg  = d & 0xFF00;
+
+	const uint32_t srcrb = s & 0xFF00FF;
+	const uint32_t srcg  = s & 0xFF00;
+
+	uint32_t drb = srcrb - dstrb;
+	uint32_t dg  =  srcg - dstg;
+
+	drb *= a;
+    dg  *= a;
+	drb >>= 8;
+	dg  >>= 8;
+
+	uint32_t rb = (drb + dstrb) & 0xFF00FF;
+	uint32_t g  = (dg  + dstg) & 0xFF00;
+
+	return rb | g;
+}
+
+/*
+ * Copy icon pixels, blend with background
+ */
+void copy_with_pixel_blend(uint32_t *dst, uint32_t* src, uint32_t background)
+{
+    int i;
+    for(i=0; i < 16*16; ++i) {
+        *dst++ = pixel_blend(background,*src++);
+    }
+}
+#endif
+
 static void x_draw_title_border(Con *con, struct deco_render_params *p) {
     assert(con->parent != NULL);
 
@@ -570,6 +611,9 @@ void x_draw_decoration(Con *con) {
     }
     //DLOG("indent_level = %d, indent_mult = %d\n", indent_level, indent_mult);
     int indent_px = (indent_level * 5) * indent_mult;
+#ifdef USE_ICONS
+    if (win->icon) indent_px += 18;
+#endif
 
     int mark_width = 0;
     if (config.show_marks && !TAILQ_EMPTY(&(con->marks_head))) {
@@ -611,6 +655,40 @@ void x_draw_decoration(Con *con) {
     if (con->title_format != NULL)
         I3STRING_FREE(title);
 
+#ifdef USE_ICONS
+    /* Draw the icon */
+    if (win->icon) {
+        xcb_image_t* icon;
+
+        uint16_t width = 16;
+        uint16_t height = 16;
+        uint32_t icon_pixels[width*height];
+
+        copy_with_pixel_blend(icon_pixels, win->icon, p->color->background);
+
+        icon = xcb_image_create_native( conn,
+                width, height,
+                XCB_IMAGE_FORMAT_Z_PIXMAP,
+                root_depth,
+                NULL,
+                width*height*4,
+                (uint8_t*)icon_pixels
+                );
+
+        if (icon) {
+            int icon_offset_y = (con->deco_rect.height - 16) / 2;
+
+            xcb_image_put(conn, parent->pixmap, parent->pm_gc,
+                    icon, con->deco_rect.x + indent_px - 16 , con->deco_rect.y + icon_offset_y, 0);
+
+            xcb_image_destroy(icon);
+        }
+        else {
+            ELOG("Error creating XCB image\n");
+        }
+    }
+#endif
+
 after_title:
     x_draw_decoration_after_title(con, p);
 copy_pixmaps:
-- 
2.9.2

